File: csharp/1385.find-the-distance-value-between-two-arrays.cs
/*
 * @lc app=leetcode id=1385 lang=csharp
 *
 * [1385] Find the Distance Value Between Two Arrays
 */

using System;

// @lc code=start
public class Solution {
    public int FindTheDistanceValue(int[] arr1, int[] arr2, int d) {
        int m = arr1.Length, n = arr2.Length, i = 0, j = 0, ret = 0;
        Array.Sort(arr1);
        Array.Sort(arr2);
        while (i < m && j < n) {
            int a = arr1[i], b = arr2[j];
            if (b < a - d) {
                ++j;
            } else if (b <= a + d) {
                ++i;
            } else {
                ++i;
                ++ret;
            }
        }
        return ret + m - i;
    }
}
// @lc code=end

File: java/1385.find-the-distance-value-between-two-arrays.java
/*
 * @lc app=leetcode id=1385 lang=java
 *
 * [1385] Find the Distance Value Between Two Arrays
 */

import java.util.Arrays;

// @lc code=start

class Solution {
    public int findTheDistanceValue(int[] arr1, int[] arr2, int d) {
        int ret = 0;
        Arrays.sort(arr2);
        for (int x : arr1) {
            if (good(arr2, x - d, x + d)) {
                ++ret;
            }
        }
        return ret;
    }

    private static boolean good(int[] arr, int min, int max) {
        int l = 0, r = arr.length - 1;
        while (l <= r) {
            int m = (l + r) / 2;
            if (arr[m] >= min && arr[m] <= max) {
                return false;
            } else if (arr[m] < min) {
                l = m + 1;
            } else {
                r = m - 1;
            }
        }
        return true;
    }
}
// @lc code=end

